pkg/credentials/filecache_v2.go

```diff
@@ -0,0 +1,110 @@
+package credentials
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/spf13/afero"
+
+	"github.com/kris-nova/logger"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+)
+
+// FileCacheV2 is a file-based credentials cache for AWS credentials that can expire,
+// satisfying the aws.CredentialsProvider interface.
+// It is meant to be wrapped with aws.CredentialsCache. The cache is per profile.
+type FileCacheV2 struct {
+	provider      aws.CredentialsProvider
+	profileName   string
+	cacheFilename string
+	fs            afero.Fs
+	newFlock      FlockFunc
+	clock         Clock
+
+	creds *aws.Credentials
+	mu    sync.Mutex
+}
+
+// NewFileCacheV2 initializes the cache and returns a *FileCacheV2.
+func NewFileCacheV2(provider aws.CredentialsProvider, profileName string, fs afero.Fs, newFlock FlockFunc, clock Clock) (*FileCacheV2, error) {
+	cacheFilename, err := initializeCache(fs)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing credentials cache: %w", err)
+	}
+	return &FileCacheV2{
+		provider:      provider,
+		profileName:   profileName,
+		cacheFilename: cacheFilename,
+		fs:            fs,
+		newFlock:      newFlock,
+		clock:         clock,
+	}, nil
+}
+
+func toAWSCredentials(c cachedCredential) *aws.Credentials {
+	return &aws.Credentials{
+		AccessKeyID:     c.Credential.AccessKeyID,
+		SecretAccessKey: c.Credential.SecretAccessKey,
+		SessionToken:    c.Credential.SessionToken,
+		Source:          c.Credential.ProviderName,
+		CanExpire:       !c.Expiration.IsZero(),
+		Expires:         c.Expiration,
+	}
+}
+
+// Retrieve implements aws.CredentialsProvider.
+func (f *FileCacheV2) Retrieve(ctx context.Context) (aws.Credentials, error) {
```

Comment:
So, what exactly changed here in the API that the other cache is no longer working?

Reply:
That's evident in the function signature:

`Retrieve(ctx context.Context) (aws.Credentials, error) {`

vs

`Retrieve() (credentials.Value, error) {`

Reply:
Shouldn't this have implemented rather our own MFA token provider which should have been using a file cache in the background?

Reply:
But I guess it doesn't matter anymore.